ts/protocol/utils.ts

```diff
@@ -211,3 +215,124 @@ export class CancellablePromise<T> extends PromiseResolver<T> {
     this.reject(CancellablePromise.CANCELLED_ERROR);
   }
 }
+
+/**
+ * Error specific to timeout racer, allows to know whether the a timeout error
+ * has been raised by a specific instance of a TimeoutRacer.
+ */
+class TimeoutRacerError extends Error {
+  constructor() {
+    super('The timeout has expired.');
+  }
+}
+
+/**
+ * Interface of the timeout racer to handle a common timeout for a series of
+ * asynchronous operations.
+ */
+export interface TimeoutRacer {
+  race<R>(promise: Promise<R>): Promise<R>;
+  stop(): void;
+  rethrowTimeoutError(error: Error): void;
+  handleTimeoutError(error: Error): void;
+  dispose(): void;
+}
+
+/**
+ * Class that handles a timeout that is passed along several subsequent
+ * asynchronous operations.
+ */
+class EnabledTimeoutRacer implements TimeoutRacer {
+  private error: TimeoutRacerError = new TimeoutRacerError();
+  private timeoutPromiseResolver: TimeoutPromiseResolver<void>;
+
+  constructor(timeoutMs: number) {
+    this.timeoutPromiseResolver =
+        new TimeoutPromiseResolver<void>(this.error, timeoutMs);
+  }
+
+  /**
+   * Races the given promise with the timeout.
+   */
+  race<R>(promise: Promise<R>): Promise<R> {
+    // If the timeout promise has already been rejected, this will directly
+    // reject. The cast is safe as the timeoutPromiseResolver never resolves.
+    return Promise.race([this.timeoutPromiseResolver.promise, promise]) as
+        Promise<R>;
+  }
+
+  /**
+   * Stops the timeout. It will neither resolve nor reject.
+   */
+  stop(): void {
+    this.timeoutPromiseResolver.clear();
+  }
+
+  /**
+   * Rethrows the error given if it is the timeout racer's error.
+   */
+  rethrowTimeoutError(error: Error) {
+    if (error === this.error) {
+      throw error;
+    }
+  }
+
+  /**
+   * Rethrows anhy error given unless it is the timeout racer's error.
+   */
+  handleTimeoutError(error: Error) {
```

Comment:
This was pretty confusing for me initially.

You should call this method `rethrowUnlessTimeoutError` or something similar.